### PR TITLE
Propagate updates

### PR DESCRIPTION
**Pull request by parth21999**

### CODE REVIEW

File: update_deps/propagate_updates.ps1 [hunk 1]
```diff
@@ -1,2 +1,509 @@
...
+
+<#
+.SYNOPSIS
+
+Propagates dependency updates for git repositories.
+
+.DESCRIPTION
+
+Given a root repo and personal access tokens for Github and Azure Devops Services, this script \
+builds the dependency graph and propagates updates from the lowest level upto the \
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:12]: >upto [](http://example.com/codeflow?start=73&length=4)

up to

File: update_deps/propagate_updates.ps1 [hunk 2]
```diff
@@ -1,2 +1,509 @@
...
+
+PS> .\{PATH_TO_SCRIPT}\propagate_updates.ps1 -azure_token {token1} -github_token {token2} [-root {root_repo_url}] [-resume]
+#>
+
+
+param (
+    [string]$root="https://msazure.visualstudio.com/DefaultCollection/One/_git/Azure-MessagingStore", # url for repo upto which updates must be propagated
+    [Parameter(Mandatory=$true)][string]$github_token, # Github personal access token: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token
+    [Parameter(Mandatory=$true)][string]$azure_token, # Azure Devops Services personal access token: https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=preview-page
+    [switch]$resume = $false # flag to continue from checkpoint of previous run
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:47]: >[switch]$resume = $false # flag to continue from checkpoint of previous run [](http://example.com/codeflow?start=4&length=75)

Why is this needed?
I would think that whenever a new run is started if we go through all the repos and we find that no update needs to be done then there would be no PRs issued and the script would effectively resume where it left off.

File: update_deps/propagate_updates.ps1 [hunk 3]
```diff
@@ -1,2 +1,509 @@
...
+        }
+    }
+    Write-Host "Done updating repo $repo_name"
+}
+
+
+# iterate over all repos and update them
+function propagate-updates {
+    # if -resume flag was not set
+    if(!$resume) {
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:472]: >if(!$resume) { [](http://example.com/codeflow?start=4&length=14)

I don't think the resume is needed

File: update_deps/propagate_updates.ps1 [hunk 4]
```diff
@@ -1,2 +1,509 @@
...
+    }
+    foreach ($repo in $repo_order) {
+        # skip repos that were updated in previous run of script
+        if($resume -and ($repo -in $done_repos)) {
+            Write-Host "Skipping repo $repo"
+            continue
+        }
+        update-repo $repo
+        # record that repo has been updated
+        Add-Content -Path done.txt -Value $repo
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:504]: I think the done repos is also not needed if we do not have the resume option

File: update_deps/propagate_updates.ps1 [hunk 5]
```diff
@@ -1,2 +1,509 @@
...
+
+
+# update dependencies for given repo
+function update-repo {
+    param(
+        [string] $repo_name
+    )
+    Write-Host "`n`nUpdating repo $repo_name"
+    [string]$git_output = (update-local-repo $repo_name)
+    if($git_output.Contains("nothing to commit")) {
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:452]: I love how you broke down the code into functions and how organized it is

File: update_deps/propagate_updates.ps1 [hunk 6]
```diff
@@ -1,2 +1,509 @@
...
+}
+
+
+# update dependencies for Azure repo
+function update-repo-azure {
+    param(
+        [string] $repo_name
+    )
+    Write-Host "`nCreating work item"
+    $work_item_id = create-work-item-azure
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:412]: Frankly generating work items does not add a lot of value.
The reasoin to attach a work item is to be able to track from a commit to a PR from the PR to a work item. As in: "why did we do the change"?
Or at least that is the imagined reason why we would do it, since it is extremely rare that we trace code changes like that.

If we generate the work item the answer to that question does not exist anymore.

I think it would be better to either:
i) ask the user for the workitem id.     or
ii) allow submodule updates without specifying the work item.

**Comment by dcristoloveanu** [comment]
Waiting on the 2 changes:
- The removal of resume
- Clearing the work item creation story

**Review by dcristoloveanu — CHANGES_REQUESTED** [review]
:clock1:

**Comment by dcristoloveanu** [comment]
>         # update repo level of submodule if path from current repo is longer

I don't see any reference to a path length check. where is that?

---
Refers to: update_deps/build_graph.ps1:136 in 74645b6. [](commit_id = 74645b62f445467842e7413efff60bb84556ca08, deletion_comment = False)

File: update_deps/propagate_updates.ps1 (hunk 2 above)
Comment (parth21999) [reply update_deps/propagate_updates.ps1:47]: Removed in next iteration.

File: update_deps/propagate_updates.ps1 (hunk 5 above)
Comment (parth21999) [reply update_deps/propagate_updates.ps1:452]: Thanks!

File: update_deps/propagate_updates.ps1 (hunk 6 above)
Comment (parth21999) [reply update_deps/propagate_updates.ps1:412]: In the next iteration, user is asked to provide work item id.

**Comment by parth21999** [comment]
>         # update repo level of submodule if path from current repo is longer

The repo level is the length of the path from the root repo. Updating comment to make this clearer

---
In reply to: [964536956](https://github.com/Azure/c-build-tools/pull/94#issuecomment-964536956) [](http://example.com/codeflow?ancestors=964536956,964536956)

---
Refers to: update_deps/build_graph.ps1:136 in 74645b6. [](commit_id = 74645b62f445467842e7413efff60bb84556ca08, deletion_comment = False)

File: update_deps/build_graph.ps1 [hunk 7]
```diff
@@ -81,37 +97,44 @@ function get-submodules {
...
-        # Write-Host "Submodule $submodule_name"
         # ignore submodule if it is i $repos_to_ignore
         if ($submodule_name -in $repos_to_ignore) {
             continue
         }
+        # $level is the length of the longest path from the root to the submodule seen so far.
         $level = 0
         [void]$repo_levels.TryGetValue($submodule_name, [ref]$level)
-        # update repo level of submodule if path from current repo is longer
+        # update repo level of submodule if path from root to submodule via current repo is longer
```
Comment (dcristoloveanu) [inline update_deps/build_graph.ps1:137]: > if path from r [](http://example.com/codeflow?start=40&length=15)

path in the graph

File: update_deps/propagate_updates.ps1 [hunk 8]
```diff
@@ -1,2 +1,493 @@
...
+    $global:azure_header = @{
+        Authorization="Basic $base64_azure_pat"
+    }
+}
+
+create-header-azure $azure_token
+
+
+# create work item that will be linked to the pull request
+function create-work-item-azure {
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:266]: >create-work-item-azure  [](http://example.com/codeflow?start=9&length=23)

Let's remove it if not used

File: update_deps/propagate_updates.ps1 [hunk 9]
```diff
@@ -1,2 +1,493 @@
...
+}
+
+
+# link work item to PR for Azure repo
+function link-work-item-to-pr-azure {
+    param(
+        [string] $pr_artifact_id
+    )
+    if(!$azure_work_item){
+        Write-Error "Updating Azure repos requires providing a work item id. Provide work item id as: -azure_work_item [id]"
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:315]: >te-Error "Updating Azure repos requires providing a work item id. Provide work item id as: -azure_work_item [id]" [](http://example.com/codeflow?start=11&length=113)

Consider printing the Ids when we fail

File: update_deps/propagate_updates.ps1 [hunk 10]
```diff
@@ -1,2 +1,493 @@
...
+        exit -1
+    }
+    $request_url = 'https://dev.azure.com/msazure/One/_apis/wit/workitems/'+$azure_work_item+'?api-version=6.0'
+    $body = @(
```
Comment (dcristoloveanu) [inline update_deps/propagate_updates.ps1:319]: Consider putting here a link with the MSDN doc

**Review by dcristoloveanu — APPROVED** [review]
:shipit:
